code/ai/aiturret.cpp

```diff
@@ -1812,7 +1812,7 @@ bool turret_fire_weapon(int weapon_num, ship_subsys *turret, int parent_objnum,
 	parent_aip = &Ai_info[Ships[Objects[parent_objnum].instance].ai_index];
 	parent_ship = &Ships[Objects[parent_objnum].instance];
 	wip = get_turret_weapon_wip(&turret->weapons, weapon_num);
-	int turret_weapon_class = WEAPON_INFO_INDEX(wip);
+	int turret_weapon_class = (wip - Weapon_info.data());
```

Comment:
As I recall, that sort of pointer arithmetic is frowned upon, especially when one isn't dealing with arrays.  I wanted to remove it, and I successfully did in two places, but wasn't able to remove it in the other two places without additional refactoring.  So I just inlined it.

So, two questions: 1) is that sort of macro discouraged?  b) should the two uses be further refactored to remove the need for it?

Reply:
I don't see the point in inlining a macro that does something discouraged; if anything, it just makes it harder to remove, because you have to find those instances by hand instead of tracking down where the macro is used.

As for that kind of pointer arithmetic being frowned upon... that's because you're encouraged to use `std::distance()` instead, to the best of my knowledge (at least I've seen people recommending `std::distance(some_vector.data(), pointer_to_elem)`, which may not be the same thing as the behavior actually being encouraged).

Reply:
`std::distance()` is exactly what I would use if I had iterators, but these aren't iterators.

I'll take another look into refactoring this.  If that proves impractical, I'll just restore the macro.

Reply:
A pointer is an iterator (by definition).

Reply:
Somehow I missed the second paragraph of your previous post.  It looks like `std::distance(Weapon_info.data(), wip)` does indeed work.